design/baremetal-operator/raid-api.md

```diff
@@ -0,0 +1,198 @@
+# raid-api
+
+## Status
+
+implementable
+
+## Summary
+
+This design describes changes needed to allow RAID storage
+configuration as part of provisioning hosts.
+
+## Motivation
+
+While it is reasonable to assume that cloud instances and VMs only
+need one drive with minimal configuration, many classes of server
+hardware need to have RAID storage configured to present fast and
+reliable storage to the OS. While it is possible today to
+pre-configure the hosts, it would be more convenient for users is
+metal3 could do that work for them.
+
+### Goals
+
+1. Provide an API to specify the RAID storage configuration to use
+   when provisioning a host.
+2. Outline the controller/provisioner logic changes needed to pass the
+   instructions to Ironic to have the storage configured.
+3. Support basic RAID levels `0`, `1`, and `1+0`.
+
+### Non-Goals
+
+1. Provide default RAID configuration.
+2. Retain existing RAID configuration on a host if it matches the
+   desired configuration at the time of provisioning (we make no
+   assumptions about preserving the state of the storage system or its
+   contents).
+3. Re-provision a host if the RAID settings change. (Users are
+   expected to explicitly deprovision the host and then reprovision
+   it.)
+4. Support every possible RAID configuration.
+
+## Proposal
+
+### User Stories
+
+1. As a user, I can specify the RAID configuration for a host before
+   provisioning it.
+2. As a user, I can choose between hardware and software RAID for each
+   host.
+
+## Design Details
+
+### API Changes
+
+The `BareMetalHost` API will change to accept RAID configuration
+inputs and to reflect the settings that were used when a host is
+provisioned (in case they are changed after provisioning).
+
+New `spec` fields will be added in a section called `raid`, with
+separate fields describing the hardware and software RAID settings. If
+the hardware settings are provided, the software settings are ignored.
```

Comment:
Then this begs the question, do we want this in BMO or not, and ensure this is implemented accordingly, without _presuming_ Ironic to do anything?

Reply:
Yes, good question. This came from the proposed implementation, so I assume it was backed by a limitation of Ironic.

What does it mean to ask for both hardware and software RAID?

Reply:
I believe this question is easier answered if we think of it as ``is there a _desired_ situation/use-case where we'd like to have both enabled at the same time``? If this isn't the case. It's safe to enforce this, to avoid ambiguous behavior. 

Whether a limitation from Ironic enforces this will be secondary if we don't need this in BMO.

Reply:
I think wanting both hardware and software RAID is a pretty exotic case that we don't want to support (just imagine recovering failed RAID in this case).

Reply:
I agree. I suggest we "discourage" this and leave it for the situation that (hopefully never) arises.

Reply:
OK, I think we agree on the limitation as stated in the doc so I am going to resolve this thread.